### PR TITLE
Update shallow_copy_and_detach for nested tensor impls to enable nested tensor softmax backward

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
This change fixes a bug that was encountered when trying to add more backward formulas for nested tensor ops.  If a derivative is defined that stores the "result" for use in the backward the output of the forward op is saved using:
```
 if (grad_fn) {
    grad_fn->result_ = SavedVariable(result, true);
  }
```

SavedVariable calls a series of functions which in turn calls shallow_copy_and_detach and when https://github.com/pytorch/pytorch/blob/c179597753f20089056115542d3b9c74ab8b864c/c10/core/TensorImpl.cpp#L533 is hit this calls sizes_custom() which is not implemented and errors.  I also noticed that since the storage format is different for nested_tensor not `storage_ ` but instead two tensors that the we should actually be calling the NestedTensorImpl constructor.

This PR overrides shallow_copy_and_detach from the derived class and ensures that shallow copy works correctly.

## Update
- Added the softmax derivative in this PR because that is a direct use case that was blocked by not having shallow_copy_and_detach work correctly.

### CODE REVIEW

**Comment by cpuhrsch** [comment]
Let's merge it into the softmax work so that we can land it with an application and a test that uses the code.

**Comment by drisspg** [comment]
Added the softmax backward formula and a grad_test check. Also added a new NestedTensorBackward.cpp file since alot of the backward formulas ops need to be registered with the dispatcher I think having this file cleans up some of the organization

**Comment by drisspg** [comment]
Will make a follow up PR to move all the factory function derivatives but to keep this one cleaner I though it okay to leave as is for now

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -890,6 +888,21 @@ def grad_test_func(a, b, c, weight, bias=None):
...
+
+        def grad_test_func(a, b, c, dim):
+            nt = torch.nested_tensor([a, b, c])
+            # This implicitly tests to_padded_tensor grads
+            d = torch.functional.F.softmax(nt, dim=dim)
+            return d.to_padded_tensor(0)
+
+        # softmax over last dim
+        data = (a, b, c, -1)
+        assert torch.autograd.gradcheck(grad_test_func, inputs=data)
```
Comment (soulitzer) [inline test/test_nestedtensor.py:904]: nit: use `gradcheck` from `torch.testing._internal.common_utils` which is basically just a wrapper around gradcheck that sets some defaults that we use in the test suite.

File: test/test_nestedtensor.py [hunk 2]
```diff
@@ -890,6 +888,21 @@ def grad_test_func(a, b, c, weight, bias=None):
         data = (a, b, c, weight)
         assert torch.autograd.gradcheck(grad_test_func, inputs=data)
```
Comment (soulitzer) [inline test/test_nestedtensor.py:889]: same here

File: aten/src/ATen/NestedTensorImpl.cpp [hunk 3]
```diff
@@ -162,5 +162,44 @@ const char* NestedTensorImpl::tensorimpl_type_name() const {
...
+    // otherwise just copy the TensorImpl and not the PyObject.  Since
+    // the interpreter is dead no one can call us out on it
+  }
+  auto impl = c10::make_intrusive<NestedTensorImpl>(this -> buffer_, this -> nested_size_tensor_);
+  copy_tensor_metadata(
+      /*src_impl=*/this,
+      /*dest_impl=*/impl.get(),
+      /*version_counter=*/std::forward<VariableVersion>(version_counter),
+      /*allow_tensor_metadata_change=*/allow_tensor_metadata_change);
+  return impl;
```
Comment (soulitzer) [inline aten/src/ATen/NestedTensorImpl.cpp:187]: This just removes the refresh numel and contiguous right?

File: aten/src/ATen/native/nested/NestedTensorMath.h [hunk 4]
```diff
@@ -14,6 +14,68 @@ int64_t get_consistent_last_dim_of_nested_tensor(const NestedTensorImpl& nt);
...
+    return shapes;
+  }
+  const Tensor& sizemat = self_ptr->get_nested_size_tensor();
+  int64_t orig_dim = sizemat.size(1);
+  // nesting scalars has empty shapes
+  if (orig_dim == 0) {
+    return shapes;
+  }
+  const int64_t* sizemat_ptr = sizemat.data_ptr<int64_t>();
+  for (int64_t i = 0; i < ntensors; i++) {
```
Comment (soulitzer) [inline aten/src/ATen/native/nested/NestedTensorMath.h:67]: nit: use irange https://github.com/pytorch/pytorch/issues?page=1&q=irange

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -4382,9 +4382,6 @@
 # softmax allows positional dtype, unlike most operators, because kwonly is BC-breaking when loading jit models.
 - func: softmax.int(Tensor self, int dim, ScalarType? dtype=None) -> Tensor
   variants: function, method
-  dispatch:
-    CompositeImplicitAutograd: softmax
-    NestedTensorCPU, NestedTensorCUDA: softmax
```
Comment (soulitzer) [inline aten/src/ATen/native/native_functions.yaml:4387]: If this gets removed, do we still go into the nested-specific softmax kernel?

hmm I also see that it was previously just calling `softmax` lol

**Comment by soulitzer** [comment]
Looks pretty good, just have some small questions

File: test/test_nestedtensor.py (hunk 1 above)
Comment (drisspg) [reply test/test_nestedtensor.py:904]: Just tried this and that one internally calls torch.stack() which is not a registered op for nested tensors

File: aten/src/ATen/NestedTensorImpl.cpp (hunk 3 above)
Comment (drisspg) [reply aten/src/ATen/NestedTensorImpl.cpp:187]: yeah exactly since those aren't supported for nested tensors

File: aten/src/ATen/native/nested/NestedTensorMath.h (hunk 4 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/NestedTensorMath.h:67]: yeah will do I was just moving code around but thats a good call out

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:4387]: So this top-level softmax.int is essentially a thin wrapper that calls _softmax and that has the mutliple dispatches. I removed this top level dispatch and then added just the _softmax kernel. The problem was that NestedTensorCPu wasn't behaving well with composite implicit kernels but a change I just made should fix this

**Comment by drisspg** [comment]
TBH I don't know why the one lightweight dispatch build is failing. I added a new file in `/native/nested/*.cpp` and looking at the GlOB in cmakelists.txt I don't know why the two native functions defined in there are not being picked up.

**Comment by soulitzer** [comment]
btw for the failure after using the internal wrapper, is that just the batched gradient tests failing? We could probably just disable that with `check_batched_grad=False`  (and add a comment about stack).

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (soulitzer) [reply aten/src/ATen/native/native_functions.yaml:4390]: Ah, makes sense!

**Comment by soulitzer** [comment]
> TBH I don't know why the one lightweight dispatch build is failing. I added a new file in /native/nested/*.cpp and looking at the GlOB in cmakelists.txt I don't know why the two native functions defined in there are not being picked up.

Looks like its failing because `BUILD_LITE_INTERPRETER=1` is enabled and we go through the first path:
https://github.com/pytorch/pytorch/blob/a46ac94677415bf2588f9635e17a3d533abacb51/aten/src/ATen/CMakeLists.txt#L169-L183

**Comment by drisspg** [comment]
> cpu_kernel_cpp

That is what I thought as well but I build locally with BUILD_LITE_INTERPRETER=1 and printed out with cmake what files made up `all_cpu_cpp` and it indeed have NestedTensorBackward.cpp in sources

**Comment by drisspg** [comment]
> btw for the failure after using the internal wrapper, is that just the batched gradient tests failing? We could probably just disable that with `check_batched_grad=False` (and add a comment about stack).

Also yup you were right made the change and added the comment

**Comment by drisspg** [comment]
> > cpu_kernel_cpp
> 
> That is what I thought as well but I build locally with BUILD_LITE_INTERPRETER=1 and printed out with cmake what files made up `all_cpu_cpp` and it indeed have NestedTensorBackward.cpp in sources

Actually was missing when flag when me and brian tried, so I am retrying with extra cmake arg and seeing what ninja.build produces for source files

**Comment by drisspg** [comment]
I think I found it.. we have a custom cmake function called `append_filelist`. This cmake function reads the bzl file and adds sources to cmake.  This I think should be commented better or documented somewhere or something lol Idk. But it is very surprising that one build system directly impacts the other so directly. The real problem may be that I just messed up and didn't think about all the build systems that need updating.

**Comment by albanD** [comment]
> But it is very surprising that one build system directly impacts the other so directly.

For context, the goal is to avoid duplication of configuration for the different build systems (including internal ones).
So changing the `build_variables.bzl` file is all you need to do to make sure that all the build systems will pick up your new file properly.

**Comment by drisspg** [comment]
> For context, the goal is to avoid duplication of configuration for the different build systems (including internal ones)

Yeah I get that and in hindsight it makes the build system as DRY as possible which is great. I guess my frustration is as a newish contributor how do I learn this information about the build system more efficiently. I ideally I wouldn't have to bug you or Nikita. I think I checked all the Wikkis and read all the documentation. PyTorch is big and information get spread out among people, maybe that's just how it goes. But #TIL lol

**Comment by albanD** [comment]
If you want to write down some of your findings in a wiki on github I would be happy to proof read it :)

**Comment by drisspg** [comment]
> If you want to write down some of your findings in a wiki on github I would be happy to proof read it :) 

Great idea I will do that!

File: aten/src/ATen/native/nested/NestedTensorBackward.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  //  Get the info about the grad
+  const Tensor &grad_sizemat = grad_ptr->get_nested_size_tensor();
+
+  TORCH_INTERNAL_ASSERT(output_sizemat.equal(grad_sizemat));
+  Tensor grad_output =
+      wrap_buffer(at::empty_like(output_buffer), output_sizemat.clone());
+
+  // split buffer into original tensors
+  std::vector<int64_t> offsets = NestedTensor_get_offsets(output_ptr);
+  std::vector<IntArrayRef> shapes = NestedTensor_get_shapes(output_ptr);
```
Comment (soulitzer) [inline aten/src/ATen/native/nested/NestedTensorBackward.cpp:80]: nit: unused var

File: aten/src/ATen/native/nested/NestedTensorMath.h [hunk 7]
```diff
@@ -14,6 +14,122 @@ int64_t get_consistent_last_dim_of_nested_tensor(const NestedTensorImpl& nt);
...
 
+at::Tensor wrap_buffer(at::Tensor buffer, at::Tensor nested_size_tensor, at::Tensor nested_stride_tensor, const std::vector<int64_t>& offsets);
+
+inline const at::Tensor& get_buffer(const at::Tensor& tensor) {
+  return get_nested_tensor_impl(tensor)->get_buffer();
+}
+
+// The starting positions of the underlying tensors in contiguous buffer memory
+// i.e. the buffer memory offsets to get the underlying tensors
+inline std::vector<int64_t> NestedTensor_get_offsets(const NestedTensorImpl* self_ptr) {
```
Comment (soulitzer) [inline aten/src/ATen/native/nested/NestedTensorMath.h:25]: hmm none of the functions in this file are new right? I don't see where they are moved from though in this PR

File: c10/core/DispatchKeySet.cpp [hunk 8]
```diff
@@ -45,8 +45,14 @@ bool isBackendDispatchKey(DispatchKey t) {
...
-constexpr DispatchKeySet math_dispatch_keyset =
-    backend_dispatch_keyset | autograd_dispatch_keyset;
+constexpr DispatchKeySet math_dispatch_keyset = backend_dispatch_keyset |
+    autograd_dispatch_keyset
+    // See Note [NestedTensor Not Included in Backend Keys]
+    // The caveat to that note is that nested_tensor is a special case
+    // where we would like to support composite implict kernels but not
+    // explicit kernels therefore we manually add the key to the
+    // math_dispatch_keyset
+    | DispatchKeySet{DispatchKey::NestedTensor};
```
Comment (soulitzer) [inline c10/core/DispatchKeySet.cpp:55]: Is this so that during key registration time, the NestedTensor entry in the keyset is also set to the ComplicitImplicitAutograd kernel registered to that op?

File: build_variables.bzl [hunk 9]
```diff
@@ -1363,6 +1363,7 @@ aten_native_source_non_codegen_list = [
     "aten/src/ATen/native/mkl/SpectralOps.cpp",
     "aten/src/ATen/native/nested/NestedTensorMath.cpp",
     "aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp",
+    "aten/src/ATen/native/nested/NestedTensorBackward.cpp",
```
Comment (soulitzer) [inline build_variables.bzl:1366]: nice!

File: aten/src/ATen/native/nested/NestedTensorMath.h (hunk 7 above)
Comment (soulitzer) [reply aten/src/ATen/native/nested/NestedTensorMath.h:25]: might've been cleaner to separate out the refactor that moves things around without changing behavior into a separate PR in the stack

File: c10/core/DispatchKeySet.cpp (hunk 8 above)
Comment (drisspg) [reply c10/core/DispatchKeySet.cpp:55]: Yeah exactly

File: aten/src/ATen/native/nested/NestedTensorMath.h (hunk 7 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/NestedTensorMath.h:25]: This is true but also I am woking on the cluster and splitting up prs is a pain and if I didn't have to I would be super down with not splitting this PR

Comment (drisspg) [reply aten/src/ATen/native/nested/NestedTensorMath.h:25]: The reason for the move though was that I added NestedTensorBackward function.. ohh wait you are right I needed them originally for the backward but I rebased and changed my impl of the backward formula so you are right I can move this back... that being said I still think it makes more sense having this inline functions live in the header file incase other files would like to use them.

Comment (soulitzer) [reply aten/src/ATen/native/nested/NestedTensorMath.h:25]: I see, keeping it here in the header is probably fine, but I don't think you removed it from the other file that you moved it from? I see 200 lines added, but softmax + shallow_copy_and_detach changes + testing only account for like 100ish.

Comment (drisspg) [reply aten/src/ATen/native/nested/NestedTensorMath.h:25]: Ohhhh you are right, I rebased my changes on to what Yifen did and he renamed some of the functions. The Get_nested_shapes should no longer be needed let me remove those

File: test/test_nestedtensor.py [hunk 10]
```diff
@@ -1084,11 +1086,26 @@ def grad_test_func(a, b, c, weight, bias=None):
...
+        assert gradcheck(grad_test_func, inputs=data, check_batched_grad=False)
 
         # Test linear with no bias added
         data = (a, b, c, weight)
-        assert torch.autograd.gradcheck(grad_test_func, inputs=data)
+        assert gradcheck(grad_test_func, inputs=data, check_batched_grad=False)
+
+    def test_nested_tensor_softmax(self):
+        a = torch.randn(1, 2, requires_grad=True, dtype=torch.float64)
+        b = torch.randn(2, 2, requires_grad=True, dtype=torch.float64)
```
Comment (cpuhrsch) [inline test/test_nestedtensor.py:1097]: These shapes could be a big bigger to make sure we're not just relying on an edge case

**Review by soulitzer — APPROVED** [review]
LGTM

**Comment by osalpekar** [comment]
@drisspg I'll need to revert this in order to revert https://github.com/pytorch/pytorch/pull/80981 cleanly. https://github.com/pytorch/pytorch/pull/80981 caused a GPU Inference SEV internally (see S283291), so we will need to revert. You should be able to rebase and then reland afterwards, since the breakage isn't related to your PR.

File: aten/src/ATen/NestedTensorImpl.cpp [hunk 11]
```diff
@@ -209,5 +209,44 @@ const char* NestedTensorImpl::tensorimpl_type_name() const {
...
+    auto r = pyobj_interpreter_.load(std::memory_order_acquire)->detach(this);
+    if (r) {
+      r->set_version_counter(std::forward<VariableVersion>(version_counter));
+      r->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
+      return r;
+    }
+    // otherwise just copy the TensorImpl and not the PyObject.  Since
+    // the interpreter is dead no one can call us out on it
+  }
+  auto impl = c10::make_intrusive<NestedTensorImpl>(this -> buffer_, this -> nested_size_tensor_);
```
Comment (drisspg) [inline aten/src/ATen/NestedTensorImpl.cpp:228]: create constructor to pass in key set

File: aten/src/ATen/NestedTensorImpl.cpp [hunk 12]
```diff
@@ -209,5 +241,44 @@ const char* NestedTensorImpl::tensorimpl_type_name() const {
...
```
Comment (drisspg) [inline aten/src/ATen/NestedTensorImpl.cpp:260]: Since softmax_backward uses SavedVariable() `shallow_copy_and_detach` gets called. The base class`TensorImpl` calls `impl->refresh_numel();` which errors for nested_tensors. Thefore we override this method and instead construct a `NestedTensorImpl` and remove the `refresh_numel` and `refresh_contiguous` call.

This PR also introduces a NestedTensorImpl constructor which accepts a key_set and sets the `ImplType` to `c10::TensorImpl::VIEW` .  I don't think though that the above constructor needs to be changed. `copy_tensor_metadata` will copy over the key_set and all relevant metadata that gets "lost" in the above constructor. The only difference is that the `ImplType` will not be set.  However, from looking at the base class shallow_copy method it does not appear that this is required. 

@albanD @soulitzer  feel free to correct me

File: c10/core/TensorImpl.cpp [hunk 13]
```diff
@@ -122,12 +122,34 @@ TensorImpl::TensorImpl(
...
+    version_counter_ = VariableVersion(/*version=*/0);
+  }
+}
+
+TensorImpl::TensorImpl(
+    DispatchKeySet key_set,
+    const caffe2::TypeMeta data_type,
+    c10::optional<c10::Device> device_opt)
+    // NOLINTNEXTLINE(performance-move-const-arg)
+    : TensorImpl(Storage{}, key_set, data_type, std::move(device_opt)) {}
```
Comment (drisspg) [inline c10/core/TensorImpl.cpp:152]: The change to this constructor was calling
going from `TensorImpl({}, key_set, data_type, std::move(device_opt)) {}` -> `TensorImpl(Storage{}, key_set, data_type, std::move(device_opt)) {}`

There was ambiguity on the dtype of the first member and this was call the correct constructor.

File: c10/core/TensorImpl.cpp [hunk 14]
```diff
@@ -122,12 +122,34 @@ TensorImpl::TensorImpl(
...
 }
 
+// This is currently used to construct view tensors for nested_tensor impls
 TensorImpl::TensorImpl(
+    ImplType type,
     DispatchKeySet key_set,
     const caffe2::TypeMeta data_type,
     c10::optional<c10::Device> device_opt)
     // NOLINTNEXTLINE(performance-move-const-arg)
-    : TensorImpl({}, key_set, data_type, std::move(device_opt)) {}
```
Comment (drisspg) [inline c10/core/TensorImpl.cpp:130]: couldn't reuse the constructor on line: 105 because this takes the device type from storage.. sigh.. so created a new constructor

File: aten/src/ATen/NestedTensorImpl.cpp (hunk 12 above)
Comment (albanD) [reply aten/src/ATen/NestedTensorImpl.cpp:260]: Don't you need to copy the NestedTensorImpl specific fields here? copy_tensor_metadata is only copying the metadata from the base TensorImpl no?

File: aten/src/ATen/native/native_functions.yaml [hunk 15]
```diff
@@ -4579,6 +4576,8 @@
 
 - func: _softmax_backward_data(Tensor grad_output, Tensor output, int dim, ScalarType input_dtype) -> Tensor
   structured_delegate: _softmax_backward_data.out
+  dispatch:
+    NestedTensorCPU, NestedTensorCUDA: nested_softmax_backward
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:4580]: You need to leave the CompositeImplicitAutograd key here

File: c10/core/TensorImpl.h [hunk 16]
```diff
@@ -477,7 +477,12 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
       Storage&& storage,
       DispatchKeySet,
       const caffe2::TypeMeta data_type);
-
+  // Used to construct a nested_tensor view
+  TensorImpl(
```
Comment (albanD) [inline c10/core/TensorImpl.h:481]: cc @ezyang about adding new constructor here?

Comment (ezyang) [reply c10/core/TensorImpl.h:481]: It's probably the wrong thing to do. I'll take a read

File: c10/core/TensorImpl.cpp (hunk 14 above)
Comment (ezyang) [reply c10/core/TensorImpl.cpp:130]: You should use this overload directly

```
TensorImpl::TensorImpl(
    Storage&& storage,
    DispatchKeySet key_set,
    const caffe2::TypeMeta data_type,
    c10::optional<c10::Device> device_opt)
```

If you want to be a good citizen, you should refactor 

```
TensorImpl::TensorImpl(
    ImplType type,
    Storage&& storage,
    DispatchKeySet key_set,
    const caffe2::TypeMeta data_type)
```

to call into the above constructor too.

File: aten/src/ATen/NestedTensorImpl.cpp (hunk 12 above)
Comment (drisspg) [reply aten/src/ATen/NestedTensorImpl.cpp:260]: yeah good call out, in the contiguous case the constructor used would create the required fields but should be directly copying all the nested specific meta data.

File: aten/src/ATen/native/native_functions.yaml (hunk 15 above)
Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:4580]: thanks good catch

File: c10/core/TensorImpl.cpp (hunk 14 above)
Comment (drisspg) [reply c10/core/TensorImpl.cpp:130]: With the first overload you mentioned though can't I not set the ImplType::VIEW enum though? I think the problem is that the body of the first construct does alot of extra work like removing keys when in inference mode and the second constructor is used to avoid all the extra work. Maybe this is a c++ thing I am confused on

Comment (ezyang) [reply c10/core/TensorImpl.cpp:130]: There is an efficiency thing going on here, but it's not relevant to your PR.

This constructor, as ImplType enum suggests, is specifically for constructing dense-to-dense views.

```
  // Note [Enum ImplType]
  // This enum is temporary. In the followup refactor we should
  // think about how to specialize TensorImpl creation for view
  // tensors. Currently we only special case its key_set_ but
  // there's also potential to share version_counter_ directly
  // without creating first and then override in as_view.
  enum ImplType { VIEW };
```

It came out of inference mode related changes, see #54403. The primary performance implication is not about skipping dispatch key manipulation (cheap) but avoiding dynamic allocations for view counters. I suppose I would not suggest you attempt this refactor, though there is a refactor opportunity as the comment suggests.

In any case, you shouldn't be touching any of this machinery in your PR. The view constructor was specifically designed storage-based tensors (which nested tensor is not), and if you want to generalize it for nested tensor there needs to be a design. But I don't think any of this matters. Just use the stock non-view constructor.

**Comment by drisspg** [comment]
This PR is doing two things and I think that is making it confusing/bad. 
Thing 

1.) Add softmax backward. Currently SavedVarible which is used to save the input for SoftmaxBackward is broken nested_tensors. One part of this PR is overriding the relevant functions in NestedTensorImpl and adding the derivative for nested_tensor softmax in order to test.

2.) A bug found that creating a base nested_tensor out of inference mode and creating a view in inference mode.  The tensor_impl constructor was removing keys on the newly constructed view of the base.  The change here was to create a new function that can create a new nested_tensor that is a "view" of a base nested_tensor. Sharing the same buffer but having different nested_sizes, nested_strides, and offsets. More though though needs to be put into the version_counter() tracking. Maybe since there aren't many/ if any inplace ops currently supported on nested_tensor this problem hasnt' arisen.

**Comment by drisspg** [comment]
I split this PR into two: This commit is isolated to changes needed to make shallow_copy_and_detach work. As well as add the backward formula for softmax as a test of these changes. 

The second PR is: https://github.com/pytorch/pytorch/pull/82658. This PR has all the changes related to creating views of nested tensors. A more formal design doc needs to be created in order to asses if #82658 encapsulates the correct design.

File: aten/src/ATen/native/native_functions.yaml (hunk 15 above)
Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:4581]: I added the key and am getting undefined symbol errors. I need to check and see if structured delegates stamp out the _softmax_backward_data() compostiteImplicit kernels.. I would think so but something seems to not be working here

File: aten/src/ATen/NestedTensorImpl.cpp [hunk 17]
```diff
@@ -108,43 +108,62 @@ c10::DispatchKeySet generate_nested_key_set(at::Tensor buffer) {
 }
 
 NestedTensorImpl::NestedTensorImpl(
-    at::Tensor buffer,
+    int64_t buffer_size,
+    Storage storage,
```
Comment (drisspg) [inline aten/src/ATen/NestedTensorImpl.cpp:112]: It has been a long day of l-value vs r-value semantics for me I think this right but honestly little confused. If someone who has a good grasp on all the const l value refs vs pass by value vs pass by r-value stuff and how it applies to our particular class types e.g. Tensor (which is a pointer class to tensor impl) and Storage (which is pointer class to storage impl) I would be eternally grateful and indebted to you for life  if you would vc me and ELI5 lol

File: aten/src/ATen/native/native_functions.yaml (hunk 15 above)
Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:4581]: Still can't get this to work if I add the CompositeImplicitAutograd. I think because it is a structured delegate the dispatch table is in fact populated from the kernels defined from here:
```
- func: _softmax_backward_data.out(Tensor grad_output, Tensor output, int dim, ScalarType input_dtype, *, Tensor(a!) grad_input) -> Tensor(a!)
  structured: True
  dispatch:
    CPU: softmax_backward_cpu_out
    CUDA: softmax_backward_cuda_out
    MPS: softmax_backward_mps_out
```
so there is no `_softmax_backward_data` compositeImpllict kernel? or at least I am not sure what the name of that composite kernel is if it does exist, because it does not match the name `_softmax_backward_data`

File: aten/src/ATen/NestedTensorImpl.cpp (hunk 17 above)
Comment (ezyang) [reply aten/src/ATen/NestedTensorImpl.cpp:112]: The rule is pretty simple: if you *always* take ownership of the passed in object, then you should probably take it by value. This is definitely true for smart pointer types. Scott Meyers has a good discussion about this, see Item 41 in https://moodle.ufsc.br/pluginfile.php/2377667/mod_resource/content/0/Effective_Modern_C__.pdf

**Comment by ezyang** [comment]
don't forget to undo the submod updates. Not doing a full review unless you ask.

**Comment by drisspg** [comment]
Because of divergence between internal diff and this one and the revert that messed things up I created a new PR: https://github.com/pytorch/pytorch/pull/83002
